File: backend/app/models/story_models.py
"""
Story Data Models - Pydantic models for story state and components
Defines the structure for story segments, choices, and overall state
"""

from typing import List, Dict, Optional
from dataclasses import dataclass
from datetime import datetime
from pydantic import BaseModel, Field


class StoryChoice(BaseModel):
    """Represents a choice option in the story"""
    text: str = Field(..., description="The choice text displayed to the user")
    action: str = Field(..., description="The action identifier for this choice")
    description: Optional[str] = Field(None, description="Additional description of the choice")
    
    class Config:
        schema_extra = {
            "example": {
                "text": "Enter the caverns boldly",
                "action": "bold_entrance",
                "description": "Charge forward with confidence"
            }
        }


class StorySegment(BaseModel):
    """Represents a segment of the story"""
    id: str = Field(..., description="Unique identifier for this segment")
    text: str = Field(..., description="The story text for this segment")
    choices: List[StoryChoice] = Field(..., description="Available choices for the user")
    background_context: Optional[str] = Field(None, description="Background context for this segment")
    mood: Optional[str] = Field(None, description="The mood/atmosphere of this segment")
    location: Optional[str] = Field(None, description="The location where this segment takes place")
    
    class Config:
        schema_extra = {
            "example": {
                "id": "crystal_caverns_entrance",
                "text": "You stand at the entrance of the ancient Crystal Caverns...",
                "choices": [
                    {"text": "Enter boldly", "action": "bold_entrance"},
                    {"text": "Study first", "action": "cautious_approach"}
                ],
                "background_context": "The Crystal Caverns are legendary...",
                "mood": "mysterious",
                "location": "Crystal Caverns Entrance"
            }
        }


class StoryState(BaseModel):
    """Represents the complete state of a story session"""
    story_id: str = Field(..., description="Unique identifier for the story session")
    current_segment: StorySegment = Field(..., description="The current story segment")
    story_history: List[StorySegment] = Field(default_factory=list, description="Previous story segments")
    character_info: Dict[str, str] = Field(default_factory=dict, description="Information about characters")
    world_info: Dict[str, str] = Field(default_factory=dict, description="Information about the world")
    genre: str = Field(..., description="The genre of the story")
    difficulty: str = Field(..., description="The difficulty level")
    created_at: datetime = Field(..., description="When the story was created")
    last_updated: Optional[datetime] = Field(None, description="When the story was last updated")
    
    class Config:
        schema_extra = {
            "example": {
                "story_id": "uuid-12345",
                "current_segment": {
                    "id": "start",
                    "text": "You begin your adventure...",
                    "choices": []
                },
                "story_history": [],
                "character_info": {"protagonist": "Brave adventurer"},
                "world_info": {"setting": "Fantasy realm"},
                "genre": "fantasy",
                "difficulty": "medium",
                "created_at": "2024-01-01T00:00:00Z"
            }
        }


class CreateStoryRequest(BaseModel):
    """Request model for creating a new story"""
    genre: str = Field(default="fantasy", description="The genre of the story")
    difficulty: str = Field(default="medium", description="The difficulty level")
    
    class Config:
        schema_extra = {
            "example": {
                "genre": "fantasy",
                "difficulty": "medium"
            }
        }


class MakeChoiceRequest(BaseModel):
    """Request model for making a story choice"""
    choice_index: int = Field(..., ge=0, description="Index of the chosen option")
    
    class Config:
        schema_extra = {
            "example": {
                "choice_index": 0
            }
        }


class StoryResponse(BaseModel):
    """Response model for story operations"""
    success: bool = Field(..., description="Whether the operation was successful")
    message: str = Field(..., description="Response message")
    data: Optional[Dict] = Field(None, description="Response data")
    
    class Config:
        schema_extra = {
            "example": {
                "success": True,
                "message": "Story created successfully",
                "data": {"session_id": "uuid-12345"}
            }
        }


class StorySegmentResponse(BaseModel):
    """Response model for story segment generation"""
    segment: StorySegment = Field(..., description="The generated story segment")
    session_id: str = Field(..., description="The story session ID")
    
    class Config:
        schema_extra = {
            "example": {
                "segment": {
                    "id": "next_segment",
                    "text": "You continue your journey...",
                    "choices": []
                },
                "session_id": "uuid-12345"
            }
        }

File: ios/NarrativeForge/Models/StoryModels.swift
import Foundation

// MARK: - Story Models

struct StoryChoice: Codable, Identifiable {
    let id = UUID()
    let text: String
    let action: String
    let description: String?
    
    enum CodingKeys: String, CodingKey {
        case text, action, description
    }
}

struct StorySegment: Codable, Identifiable {
    let id: String
    let text: String
    let choices: [StoryChoice]
    let backgroundContext: String?
    let mood: String?
    let location: String?
    
    enum CodingKeys: String, CodingKey {
        case id, text, choices, backgroundContext, mood, location
    }
}

struct StoryState: Codable {
    let storyId: String
    let currentSegment: StorySegment
    let storyHistory: [StorySegment]
    let characterInfo: [String: String]
    let worldInfo: [String: String]
    let genre: String
    let difficulty: String
    let createdAt: String
    let lastUpdated: String?
    
    enum CodingKeys: String, CodingKey {
        case storyId = "story_id"
        case currentSegment = "current_segment"
        case storyHistory = "story_history"
        case characterInfo = "character_info"
        case worldInfo = "world_info"
        case genre, difficulty
        case createdAt = "created_at"
        case lastUpdated = "last_updated"
    }
}

// MARK: - API Request Models

struct CreateStoryRequest: Codable {
    let genre: String
    let difficulty: String
}

struct MakeChoiceRequest: Codable {
    let choiceIndex: Int
    
    enum CodingKeys: String, CodingKey {
        case choiceIndex = "choice_index"
    }
}

// MARK: - API Response Models

struct StoryResponse: Codable {
    let success: Bool
    let message: String
    let data: [String: AnyCodable]?
}

struct StorySegmentResponse: Codable {
    let segment: StorySegment
    let sessionId: String
    
    enum CodingKeys: String, CodingKey {
        case segment
        case sessionId = "session_id"
    }
}

// MARK: - Genre and Difficulty Models

struct Genre: Codable, Identifiable {
    let id: String
    let name: String
    let description: String
}

struct Difficulty: Codable, Identifiable {
    let id: String
    let name: String
    let description: String
}

// MARK: - Helper for AnyCodable

struct AnyCodable: Codable {
    let value: Any
    
    init(_ value: Any) {
        self.value = value
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if container.decodeNil() {
            self.value = NSNull()
        } else if let bool = try? container.decode(Bool.self) {
            self.value = bool
        } else if let int = try? container.decode(Int.self) {
            self.value = int
        } else if let double = try? container.decode(Double.self) {
            self.value = double
        } else if let string = try? container.decode(String.self) {
            self.value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            self.value = array.map { $0.value }
        } else if let dictionary = try? container.decode([String: AnyCodable].self) {
            self.value = dictionary.mapValues { $0.value }
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "AnyCodable value cannot be decoded")
        }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        switch self.value {
        case is NSNull:
            try container.encodeNil()
        case let bool as Bool:
            try container.encode(bool)
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let string as String:
            try container.encode(string)
        case let array as [Any]:
            try container.encode(array.map { AnyCodable($0) })
        case let dictionary as [String: Any]:
            try container.encode(dictionary.mapValues { AnyCodable($0) })
        default:
            let context = EncodingError.Context(codingPath: container.codingPath, debugDescription: "AnyCodable value cannot be encoded")
            throw EncodingError.invalidValue(self.value, context)
        }
    }
}
